test-runner-poc/lib/runner.js

```diff
@@ -1,73 +1,140 @@
-		(!filter.filter || typeof filter.filter !== "function")
-	) {
-		throw new Error("Object filters must have a function named filter");
-	}
-
-	if (typeof filter === "function") {
-		filters.push({ filter: filter });
-	} else {
-		filters.push(filter);
-	}
+  switch (typeof filter) {
+    case 'function':
+      filters.push({ filter: filter });
+      break;
+    case 'object':
+      if (!filter.filter || typeof filter.filter !== 'function') {
+        throw new Error('Object filters must have a function named filter');
+      }
+      filters.push(filter);
+      break;
+    default:
+      throw new Error('Type of filter must either be a function or an object');
+  }
 }
 
-before(function() {
+function environmentSetup(environmentCallback) {
+  const mongodb_uri = process.env.MONGODB_URI || 'mongodb://localhost:27017';
+  mongoClient = new MongoClient(mongodb_uri);
+
+  mongoClient.connect((err, client) => {
+    if (err) throw new Error(err);
+
+    createFilters(callback);
+
+    function callback(environmentName, version) {
+      console.log('environmentName ', environmentName);
+      const Environment = environments[environmentName];
+      const environment = new Environment(version);
+      environmentName !== 'single'
+        ? parseConnectionString(mongodb_uri, (err, parsedURI) => {
+            if (err) console.log(err);
+            environment.url = mongodb_uri + '/integration_tests';
+            environment.port = parsedURI.hosts[0].port;
+            environment.host = parsedURI.hosts[0].host;
+          })
+        : undefined;
+
+      try {
+        const mongoPackage = {
+          path: path.resolve(process.cwd(), '..'),
+          package: 'mongodb'
+        };
+        environment.mongo = require(mongoPackage.path);
+      } catch (err) {
+        throw new Error('The test runner must be a dependency of mongodb or mongodb-core');
+      }
+      environmentCallback(environment, client);
+    }
+  });
+}
+
+function createFilters(callback) {
+  let filtersInitialized = 0;
+  const filterFiles = fs.readdirSync(path.join(__dirname, 'filters'));
+
+  let topology, version;
 
-	const environmentName = process.env['MONGODB_ENVIRONMENT'];
+  filterFiles.filter(x => x.indexOf('js') !== -1).forEach(x => {
+    const FilterModule = require(path.join(__dirname, 'filters', x));
+    const filter = new FilterModule();
 
-	console.log(`[environment: ${environmentName}]`);
+    if (typeof filter.initializeFilter === 'function') {
+      filter.initializeFilter(_increment);
+    } else {
+      _increment();
+    }
 
-	//apply filters
-	fs.readdirSync(path.join(__dirname, "filters"))
-		.filter(x => x.indexOf("js") !== -1)
-		.forEach(x => {
-			const FilterModule = require(path.join(__dirname, "filters", x));
-			addFilter(new FilterModule({ runtimeTopology: 'sharded' }));
-		});
+    topology = topology || filter.runtimeTopology;
+    version = version || filter.mongoVersion;
 
+    function _increment() {
```

Comment:
Can we add some documentation to explain what this is doing?

Reply:
Added documentation - let me know if I should go into more detail!

Reply:
Yeah, I think more detail would be helpful! I like the line you added so maybe you can add more description within the function? Specifically, explaining `filtersInitialized += 1` and `if (filtersInitialized === filterFiles.length)` would be helpful!